(error message) Consensus error during delete operation

if a file doesn't belong to the allocation, and when I try to delete that file I get consensus error: 

```
zbox delete --allocation $alloc --remotepath /a/go.sum --silent
Delete failed. consensus_not_met: Consensus on delete failed. Required consensus 2 got 0
```

the gosdk log is: 
```

0chain-core-sdk [INFO]   2023/05/07 21:47:20.116856 wallet_base.go:310: ******* Wallet SDK Version:v1.8.15-42-ge1ae9e37 ******* (SetLogFile)
0box-sdk        [INFO]   2023/05/07 21:47:20.117095 sdk.go:78: ******* Storage SDK Version: v1.8.15-42-ge1ae9e37 *******
0chain-core-sdk [INFO]   2023/05/07 21:47:20.177342 transaction_base.go:139: ******* Wallet SDK Version:v1.8.15-42-ge1ae9e37 ******* (InitZCNSDK)
0box-sdk        [DEBUG]  2023/05/07 21:47:20.238611 networkworker.go:105: Get network result:{"miners":["http://100.78.70.98:7071","http://100.78.70.98:7072","http://100.78.70.98:7074","http://100.78.70.98:7073"],"sharders":["http://100.78.70.98:7171","http://100.78.70.98:7172"]}
0box-sdk        [ERROR]  2023/05/07 21:47:20.484916 common.go:37: Object tree response : 400
0box-sdk        [ERROR]  2023/05/07 21:47:20.484960 common.go:37: Object tree response : 400
0box-sdk        [ERROR]  2023/05/07 21:47:20.485159 deleteworker.go:183: 400: Object tree error response: Body: {"error":"invalid_signature: Invalid signature"}
0box-sdk        [ERROR]  2023/05/07 21:47:20.485209 deleteworker.go:183: 400: Object tree error response: Body: {"error":"invalid_signature: Invalid signature"}
```


but for upload, it show the error message correctly: 
```
zbox upload --allocation $alloc --localpath ./go.sum --remotepath /new/go.sum
Upload failed. upload_failed: Upload failed. bad request: {"code":"invalid_operation","error":"invalid_operation: Operation needs to be performed by the owner or the payer of the allocation"}
```

--- a/zboxcore/sdk/deleteworker.go
+++ b/zboxcore/sdk/deleteworker.go
@@ -42,8 +42,6 @@ type DeleteRequest struct {
 func (req *DeleteRequest) deleteBlobberFile(
 	blobber *blockchain.StorageNode, blobberIdx int) error {
 
-	defer req.wg.Done()
-
 	var err error
 
 	defer func() {
@@ -161,10 +159,10 @@ func (req *DeleteRequest) ProcessDelete() (err error) {
 	var deleteMutex sync.Mutex
 	removedNum := 0
 	req.wg = &sync.WaitGroup{}
-	req.wg.Add(num)
 
 	var pos uint64
 	for i := req.deleteMask; !i.Equals64(0); i = i.And(zboxutil.NewUint128(1).Lsh(pos).Not()) {
+		req.wg.Add(1)
 		pos = uint64(i.TrailingZeros())
 		go func(blobberIdx uint64) {
 			defer req.wg.Done()
@@ -189,15 +187,13 @@ func (req *DeleteRequest) ProcessDelete() (err error) {
 	req.wg.Wait()
 
 	req.consensus.consensus = removedNum
-	numDeletes := req.deleteMask.CountOnes()
-
-	req.wg.Add(numDeletes)
 
 	var errCount int32
 	wgErrors := make(chan error)
 	wgDone := make(chan bool)
 
 	for i := req.deleteMask; !i.Equals64(0); i = i.And(zboxutil.NewUint128(1).Lsh(pos).Not()) {
+		req.wg.Add(1)
 		pos = uint64(i.TrailingZeros())
 		go func(blobberIdx uint64) {
 			defer req.wg.Done()
@@ -329,6 +325,7 @@ func (dop *DeleteOperation) Process(allocObj *Allocation, connectionID string) (
 		pos = uint64(i.TrailingZeros())
 		deleteReq.wg.Add(1)
 		go func(blobberIdx int) {
+			defer deleteReq.wg.Done()
 			refEntity, err := deleteReq.getObjectTreeFromBlobber(pos)
 			if errors.Is(err, constants.ErrNotFound) {
 				deleteReq.consensus.Done()

--- a/zboxcore/sdk/deleteworker.go
+++ b/zboxcore/sdk/deleteworker.go
@@ -338,7 +338,10 @@ func (dop *DeleteOperation) Process(allocObj *Allocation, connectionID string) (
 				l.Logger.Error(err.Error())
 				return
 			}
-			deleteReq.deleteBlobberFile(deleteReq.blobbers[blobberIdx], blobberIdx)
+			err = deleteReq.deleteBlobberFile(deleteReq.blobbers[blobberIdx], blobberIdx)
+			if err != nil {
+				blobberErrors[blobberIdx] = err
+			}
 			objectTreeRefs[blobberIdx] = refEntity
 		}(int(pos))
 	}

--- a/zboxcore/sdk/deleteworker.go
+++ b/zboxcore/sdk/deleteworker.go
@@ -7,6 +7,7 @@ import (
 	"net/http"
 	"net/url"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/0chain/errors"
@@ -39,7 +40,7 @@ type DeleteRequest struct {
 }
 
 func (req *DeleteRequest) deleteBlobberFile(
-	blobber *blockchain.StorageNode, blobberIdx int) {
+	blobber *blockchain.StorageNode, blobberIdx int) error {
 
 	defer req.wg.Done()
 
@@ -62,7 +63,7 @@ func (req *DeleteRequest) deleteBlobberFile(
 	httpreq, err := zboxutil.NewDeleteRequest(blobber.Baseurl, req.allocationTx, query)
 	if err != nil {
 		l.Logger.Error(blobber.Baseurl, "Error creating delete request", err)
-		return
+		return err
 	}
 
 	var (
@@ -123,15 +124,15 @@ func (req *DeleteRequest) deleteBlobberFile(
 		}()
 
 		if err != nil {
-			return
+			return err
 		}
 
 		if shouldContinue {
 			continue
 		}
-		return
+		return nil
 	}
-	err = errors.New("unknown_issue",
+	return errors.New("unknown_issue",
 		fmt.Sprintf("latest response code: %d", resp.StatusCode))
 }
 
@@ -192,11 +193,36 @@ func (req *DeleteRequest) ProcessDelete() (err error) {
 
 	req.wg.Add(numDeletes)
 
+	var errCount int32
+	wgErrors := make(chan error)
+	wgDone := make(chan bool)
+
 	for i := req.deleteMask; !i.Equals64(0); i = i.And(zboxutil.NewUint128(1).Lsh(pos).Not()) {
 		pos = uint64(i.TrailingZeros())
-		go req.deleteBlobberFile(req.blobbers[pos], int(pos))
+		go func(blobberIdx uint64) {
+			defer req.wg.Done()
+			err = req.deleteBlobberFile(req.blobbers[blobberIdx], int(blobberIdx))
+			if err != nil {
+				logger.Logger.Error("error during deleteBlobberFile", err)
+				errC := atomic.AddInt32(&errCount, 1)
+				if errC > int32(req.consensus.fullconsensus-req.consensus.consensusThresh) {
+					wgErrors <- err
+				}
+			}
+		}(pos)
 	}
-	req.wg.Wait()
+
+	go func() {
+		req.wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-wgErrors:
+		return thrown.New("delete_failed", fmt.Sprintf("Delete failed. %s", err.Error()))
+	}
 
 	if !req.consensus.isConsensusOk() {
 		return errors.New("consensus_not_met",
@@ -321,7 +347,7 @@ func (dop *DeleteOperation) Process(allocObj *Allocation, connectionID string) (
 	if !deleteReq.consensus.isConsensusOk() {
 		err := zboxutil.MajorError(blobberErrors)
 		if err != nil {
-			return nil, deleteReq.deleteMask, thrown.New("delete_falied", fmt.Sprintf("Delete failed. %s", err.Error()))
+			return nil, deleteReq.deleteMask, thrown.New("delete_failed", fmt.Sprintf("Delete failed. %s", err.Error()))
 		}
 
 		return nil, deleteReq.deleteMask, thrown.New("consensus_not_met",

--- a/zboxcore/sdk/deleteworker_test.go
+++ b/zboxcore/sdk/deleteworker_test.go
@@ -349,7 +349,7 @@ func TestDeleteRequest_ProcessDelete(t *testing.T) {
 			numCorrect:  2,
 			setup:       setupHttpResponses,
 			wantErr:     true,
-			errMsg:      "consensus_not_met",
+			errMsg:      "delete_failed",
 		},
 		{
 			name:        "Test_All_Blobber_Error_On_Delete_Attribute_Failure",
@@ -357,7 +357,7 @@ func TestDeleteRequest_ProcessDelete(t *testing.T) {
 			numCorrect:  0,
 			setup:       setupHttpResponses,
 			wantErr:     true,
-			errMsg:      "consensus_not_met",
+			errMsg:      "delete_failed",
 		},
 	}
 	for _, tt := range tests {

